### ISSUE TITLE
SD3 Dreambooth/Lora training scripts don't actually unload the text encoders

### ISSUE THREAD

**Issue by spacepxl** [huggingface/diffusers#8511]
### Describe the bug

In the SD3 Dreambooth training script https://github.com/huggingface/diffusers/blob/7f51f286a5397cb3e5c5a25693681aa4955e6241/examples/dreambooth/train_dreambooth_sd3.py#L1349 and similarly in the Lora training script https://github.com/huggingface/diffusers/blob/7f51f286a5397cb3e5c5a25693681aa4955e6241/examples/dreambooth/train_dreambooth_lora_sd3.py#L1318 the code deletes the *lists* of tokenizers and text encoders, but because they are referenced directly before (if I understood this correctly), the actual objects are not deleted, and thus the models remain in memory, consuming about 10 GB of VRAM unnecessarily. 

Adding an additional step of 
```
del tokenizer_one, tokenizer_two, tokenizer_three
del text_encoder_one, text_encoder_two, text_encoder_three
```
allows the models to actually be deleted and free up the memory used. There might be a more elegant solution to this, but that was my amateur fix, which reduced memory usage during the training loop from ~38 GB to ~28 GB.

### Reproduction

The bug can be observed by watching memory usage, it does not decline after the text embeddings are encoded, when the text models are supposed to be deleted. Any SD3 Dreambooth or Lora training config where the text encoders aren't being trained has this behavior.

After adding these lines of code:
```
del tokenizer_one, tokenizer_two, tokenizer_three
del text_encoder_one, text_encoder_two, text_encoder_three
```
the observed memory usage drops significantly once the text encoder models are deleted.

### System Info

- 🤗 Diffusers version: 0.29.0.dev0
- Platform: Windows-10-10.0.19045-SP0
- Running on a notebook?: No
- Running on Google Colab?: No
- Python version: 3.10.11
- PyTorch version (GPU?): 2.2.2+cu118 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.23.3
- Transformers version: 4.41.2
- Accelerate version: 0.31.0
- PEFT version: 0.11.1
- Bitsandbytes version: 0.43.1
- Safetensors version: 0.4.3
- xFormers version: 0.0.25.post1+cu118
- Accelerator: NVIDIA GeForce RTX 3090, 24576 MiB VRAM
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@sayakpaul

**Comment by haofanwang** [comment]
It's not a bug, as they are used in following code snippet.

**Comment by sayakpaul** [comment]
For the non-LoRA DreamBooth, we allow the training of text encoders but for the LoRA one, we don't (simply to keep things simple). 

So, we need to be a little careful. But in any case, could you maybe open a PR with the fix you have in mind and we can take it from there?

**Comment by spacepxl** [comment]
https://github.com/huggingface/diffusers/pull/8536

As far as I understand, the deletion and garbage collection is only triggered if the text encoders are not needed during training. @haofanwang they are referenced later in the script, but only under conditions that would not have triggered this section of the code. My change shouldn't affect text encoder training as far as I know, it is simply editing the conditional deletion to work as I believe was intended.

Side note, I've been trying to get the memory usage for DB under 24 GB. Changing the optimizer to Adafactor gets close, and I believe once the gradient accumulation bug is fixed it will be possible to get it down to about 22-23 GB. Right now with the text encoder deletion fix, and Adafactor, it's sitting at around 24.5 GB, which is functional with shared memory but slightly more than ideal.

**Comment by bghira** [comment]
i don't think running the training at such low memory usage levels will meaningfully train this particular SD3 release. do you have experimental results to show yet?

**Comment by spacepxl** [comment]
I can't share the results publicly but yes, it did learn just fine.

```python
optimizer = Adafactor(
    params_to_optimize,
    lr                = args.learning_rate, # 3e-5
    eps               = (1e-30, 1e-3),
    clip_threshold    = 1.0,
    decay_rate        = -0.8,
    beta1             = None,
    weight_decay      = 0.0,
    relative_step     = False,
    scale_parameter   = False,
    warmup_init       = False,
````
I ran batch size 1, 5 epochs on 1100 images, it needs more but was definitely learning effectively. My intuition so far is that it behaves very similarly to SDXL unet finetuning. I think once gradient checkpointing works there will be some headroom to bump up the batch size as well, which should allow for larger batch sizes and regularization, especially if the latents are precomputed and the vae offloaded. No idea about whether the hyperparameters are optimal, I've only been testing functionality so far.

**Comment by scarbain** [comment]
@spacepxl This fixes the gradient checkpointing error : https://github.com/huggingface/diffusers/pull/8542
